trusted/pal/src/pal.rs

```diff
@@ -0,0 +1,174 @@
+    ///
+    /// # Note
+    ///
+    /// The serialized representation must contain enough information to
+    /// verify the attestation itself (e.g. that it has been produced by a proper root
+    /// of trust and that it represents a trusted chain of components) and to verify
+    /// signed application data (e.g. trusted applications sharing data with each other).
+    ///
+    /// For example, before two trusted application peers can start interacting they must
+    /// establish trust. The trusted application initiating the interaction will first
+    /// obtain its own attestation, sign the data it wants to share with the peer,
+    /// share serialized attestation, the data and its signature with the peer. The peer
+    /// deserializes and verifies the attestation (see `Host::verify_peer_attestation`),
+    /// uses the attestation to verify the data signatures, performs similar handshake
+    /// with the peer.
+    fn serialize(&self) -> Result<Vec<u8>>;
+
+    /// Signs a byte string using private signing key.
+    ///
+    /// # Arguments
+    ///
+    /// * `data` - The byte string to sign.
+    ///
+    /// # Returns
+    ///
+    /// An error if the attestation doesn't possess the signing capability (e.g. this
+    /// instance has been transferred over the wire and meant to be used for verification)
+    /// or if the signing failed, a success otherwise.
+    fn sign(&self, data: &[u8]) -> Result<Vec<u8>>;
+
+    /// Verifies signature of a byte string using public signing key.
+    ///
+    /// # Arguments
+    ///
+    /// * `data` - The byte string to verify signature for.
+    /// * `signature` - The signature of the byte string.
+    ///
+    /// # Returns
+    ///
+    /// An error if the signature cannot be verifed or is invalid, a success otherwise.
+    fn verify(&self, data: &[u8], signature: &[u8]) -> Result<()>;
+
+    /// Gets serialized public key used for signing by the trusted application.
+    /// The signing key is generated at the start of the trusted application and
+    /// remains unchanged through the lifetime.
+    ///
+    /// # Note
+    ///
+    /// Public signing key can be used to derive a trusted application identity.
+    /// For example, a trusted application that represents a node in Raft cluster
+    /// running inside of a group trusted hosts, it is important that Raft node
+    /// identity cannot be forged, does not require coordination to pick one and
+    /// has low collision chance. Hash of the public signing key is an identity
+    /// mechanism that is compliant with these requirements.
+    fn public_signing_key(&self) -> Vec<u8>;
+}
+
+/// Represents a trusted host, abstracting away the details of how trusted
+/// application is hosted (e.g. trusted host can be restricted kernel bare
+/// metal based encrypted virtual machine or a linux kernel docker based
+/// encrypted virtual machine).
+pub trait Host {
```

Comment:
Maybe this should be called Platform or something like that?

Reply:
Agree, platform would be a better name here, will rename it.